core/ulog/in_memory_log.go

```diff
@@ -0,0 +1,41 @@
+package ulog
+
+import (
+	"bytes"
+	"strings"
+	"testing"
+
+	"github.com/uber-go/zap"
+)
+
+// TestBuffer is a buffer used to test the zap logger
+type TestBuffer struct {
+	bytes.Buffer
+}
+
+// Sync is a noop
+func (b *TestBuffer) Sync() error {
+	return nil
+}
+
+// Lines returns buffer as array of strings
+func (b *TestBuffer) Lines() []string {
+	output := strings.Split(b.String(), "\n")
+	return output[:len(output)-1]
+}
+
+// Stripped returns buffer as a string without the newline
+func (b *TestBuffer) Stripped() string {
+	return strings.TrimRight(b.String(), "\n")
+}
+
+// WithInMemoryLogger creates an in-memory zap logger that can be used in tests
+func WithInMemoryLogger(t *testing.T, opts []zap.Option, f func(zap.Logger, *TestBuffer)) {
```

Comment:
We can move this to testutil at some point and use in-memory logger for validating logging messages. I am ok with exporting for now, may be add a TODO?